File: kdTree.cxx
#include "kdTree.h"
#include <cstring>
#include <cstdlib>
#include <algorithm>

kdTree::~kdTree() {
    freeMemory(root);
}

void
kdTree::freeMemory(Node* parent) {

    if (parent != NULL) {
        Node* left = parent->Left;
        Node* right = parent->Right;
        if (left == NULL && right == NULL) {// At leaf the left/right nodes are null.
            delete parent;
        }
        else {
            freeMemory(left);
            freeMemory(right);
            delete parent;
        }
    }
}

void
kdTree::insert(std::vector<std::vector<double>>& kdData) {
    root = BuildkdTree(kdData, 0);
}

Node*
kdTree::BuildkdTree(std::vector<std::vector<double>>&  kdData,
                    size_t                             depth)
{
    size_t num_of_points = kdData.size();
    std::vector<std::vector<double>> P1;
    std::vector<std::vector<double>> P2;
    double median;
    size_t currentDimension;

    //Logic starts;
    if (num_of_points == 1) {
        Node* leaf = new Node();
        leaf->point = kdData[0];
        return leaf;
    }
    else if (num_of_points != 0 && num_of_points != 1) {

        currentDimension = depth % dimension;
        //sort the array.
        auto comparator = [currentDimension](std::vector<double>& a, std::vector<double>& b) {return a[currentDimension] < b[currentDimension]; };
        std::sort(kdData.begin(), kdData.end(), comparator);

        if (num_of_points % 2 == 0) {//if number of data points are even
            auto middlePoint = num_of_points / 2;
            median = (kdData[middlePoint][currentDimension] + kdData[middlePoint - 1][currentDimension]) / 2.0;
            for (auto& data : kdData) {
                if (data[currentDimension] <= median) {
                    P1.push_back(data);
                }
                else {
                    P2.push_back(data);
                }
            }
        }
        else {//if number of data points are odd
            auto middlePoint = (num_of_points - 1) / 2;
            median = kdData[middlePoint][currentDimension];
            for (auto& data : kdData) {
                if (data[currentDimension] <= median) {
                    P1.push_back(data);
                }
                else {
                    P2.push_back(data);
                }
            }
        }

        Node* left = BuildkdTree(P1, depth + 1);//bottom
        Node* right = BuildkdTree(P2, depth + 1);//top
        Node* parent = new Node(median);
        parent->Left = left;
        parent->Right = right;
        parent->splitaxis = currentDimension;
        return parent;
    }
    return NULL;
}

Node* 
kdTree::getRoot() {
    return root;
}

File: kdTree.py
import os
import statistics
import math
import numpy as np

class Node:
    def __init__(self, point=[], left=None, right=None, splitaxis=None,median=None):
        self.left = left
        self.right = right
        self.point =point
        self.splitaxis = splitaxis
        self.median = median

class kdTree:
    def __init__(self, dimension):
        self.dimension = dimension
        self.root = None

    def insert(self, points):
        self.root = self.__buildKdTree(points, 0)

    def __buildKdTree(self, points, depth):
        if not len(points):
            return None

        splittingaxis = depth % self.dimension
        if len(points) == 1:
            return Node(points[0])
        elif len(points) != 0 and len(points) != 1:
            points.sort(key=lambda point: point[splittingaxis])
            splittingList = [point[splittingaxis] for point in points]
            median = statistics.median(splittingList)
            P1 = [point for point in points if point[splittingaxis] <= median ]
            P2 = [point for point in points if point[splittingaxis] > median]
            left = self.__buildKdTree(P1,depth+1)
            right = self.__buildKdTree(P2,depth+1)
            newNode = Node([],left,right,splittingaxis,median)
            return newNode

    def getNNP(self, testPoint):
        self.nearestDistance = math.inf
        self.nearestPoint = None
        self.__searchNNP(self.root, testPoint)
        return self.nearestPoint

    def __searchNNP(self, treeNode, testPoint ):
        if treeNode:
            if not treeNode.left and not treeNode.right:
                point_a = np.array(testPoint)
                point_b = np.array(treeNode.point)
                distance = np.linalg.norm(point_a-point_b)
                if distance < self.nearestDistance:
                    self.nearestDistance = distance
                    self.nearestPoint = point_b
            else:
                if testPoint[treeNode.splitaxis] - self.nearestDistance <= treeNode.median:
                    first_left = True
                else:
                    first_left = False

                if first_left:
                    if testPoint[treeNode.splitaxis] - self.nearestDistance <= treeNode.median:
                        self.__searchNNP(treeNode.left, testPoint)
                    if testPoint[treeNode.splitaxis] + self.nearestDistance > treeNode.median:
                        self.__searchNNP(treeNode.right, testPoint)
                else:
                    if testPoint[treeNode.splitaxis] + self.nearestDistance > treeNode.median:
                        self.__searchNNP(treeNode.right, testPoint)
                    if testPoint[treeNode.splitaxis] - self.nearestDistance <= treeNode.median:
                        self.__searchNNP(treeNode.left, testPoint)
                
def testKdTree():
    reader = open('UnitSquare1000.txt', 'r')
    fileData = list(reader)
    points = [line.split(' ') for line in fileData]
    points = [[ float(coord) for coord in point ] for point in points ]
    tree = kdTree(2)
    tree.insert(points)
    nearestPoint = tree.getNNP([0.5,0.7])
    print(nearestPoint)

testKdTree()